syntax = "proto3";

package kvstore;

service KeyValueService {
  rpc Set(SetRequest) returns (SetResponse);
  rpc Get(GetRequest) returns (GetResponse);
  rpc Delete(DeleteRequest) returns (DeleteResponse);
  rpc InitialSync(SyncRequest) returns (FullSyncResponse);
  rpc SyncStream(SyncRequest) returns (stream SyncEvent);
}

service ReplicaService {
  rpc Get(GetRequest) returns (GetResponse);
}

message SyncRequest {}

message FullSyncResponse {
  repeated KVPair entries = 1;
}

message KVPair {
  string key = 1;
  oneof value {
    string s = 2;
    int32 i = 3;
    float f = 4;
    bool b = 5;
  }
}

enum OperationType {
  SET = 0;
  DELETE = 1;
}

message SyncEvent {
  OperationType op = 1;
  KVPair pair = 2;
}

message SetRequest {
  string key = 1;
  oneof value {
    string s = 2;
    int32 i = 3;
    float f = 4;
    bool b = 5;
  }
}

message SetResponse {
  bool success = 1;
}

message GetRequest {
  string key = 1;
}

message GetResponse {
  bool found = 1;
  oneof value {
    string s = 2;
    int32 i = 3;
    float f = 4;
    bool b = 5;
  }
}

message DeleteRequest {
  string key = 1;
}

message DeleteResponse {
  bool success = 1;
}
